### ISSUE TITLE
New hardware suggestion teensy3.1

### ISSUE THREAD

**Issue by bobalony1** [grbl/grbl#346]
I know that the idea for moving to new hardware has been thrown around a lot. Mostly due to the 20mhz limit. But also things like the Due's lack of eeprom have been holding new hardware back. 
My suggestion would be something like the Teensy 3.1 (also the developer is working on a ++ model to breakout the bottom side solder mount pads, so hopefully thats coming soon)

a quick list of specks taken from sparkfun
    3.3v power
    32 bit ARM Cortex-M4 72 MHz CPU (M4 = DSP extensions) (overclockable to 96mhz)
    256K Flash Memory, 64K RAM, 2K EEPROM
    21 High Resolution Analog Inputs (13 bits usable, 16 bit hardware)
    34 Digital I/O Pins (5V tolerance on Digital Inputs)
    12 PWM outputs
    7 Timers for intervals/delays, separate from PWM
    USB with dedicated DMA memory transfers
    3 UARTs (serial ports)
    SPI, I2C, I2S,CAN Bus, IR modulator
    I2S (for high quality audio interface)
    Real Time Clock (with user-added 32.768 crystal and battery)
    4 general purpose DMA channels (separate from USB)
    Touch Sensor Inputs

It uses the arduino IDE and all the libraries. and If i recall correctly I saw that someone was working on making GRBL a sketch. so that might be a nice easy way to put the two together. 

price wise its only $20 USD

dev page here
http://www.pjrc.com/teensy/teensy31.html

one downside is that it isn't open source

truth be told I don't know enough about hardware to evaluate it properly for GRBLs needs but I thought Id put it out here for you guys.

**Comment by J-Dunn** [comment]
@ langwadt , thanks.  That looks very useful. 

@bronek999 Thanks for sharing the code. Any progress since last ?!  

I was wondering about porting GRBL to M4, this should save some time. The STM32F4Discovery has an onboard QVGA touch screen, but it's not ardiuno pin layout so some prototyped, pin adapter board will be needed  Now if I can pull all these threads together it should give a cheap, super powerful CNC solution.

As soon as GRBL is on M4 a whole new world of possibilities opens up.

**Comment by timryder** [comment]
@bronek999 The link to the .zip file containing the source code is broken. 404 error. 
Can you repost or do as @chamnit suggests and create a repo on GitHub?

Amazing work guys!

**Comment by bronek999** [comment]
Hi

Version 0.8 works on my board (STM32F103+5inch LCD+SD card).
I tried port version 0.9 without success....It is 100% rewriten, 
diference between 0.8 and 0.9 on every line of code

At present I'm porting TinyG2. It is Arduino Due based (AT91SAM3S8E 
32bit MCU) board.
I need XYZ + one rotary A ....and GRBL support only 3 axes...
It is faster like GRBL, S curve accel, Jirk, Jogging supported......
Now I can read file from SD card and proccess this file.
My plan is connect 7" touch screen, I'm waiting for LCD

**Comment by J-Dunn** [comment]
Hey we're definitely on the same path.  I was thinking of porting TinyG2 but when I saw this thread here, I thought GRBL may be a good first step to get some hardware spinning. 

I'm a little wary of the complexity of the TinyG code. I could imagine it being  a nightmare to debug. 

Were there major problems with porting v0.9 , or was it just too much like starting from scratch because of all the changes?

The Tiny boards don't have enought drive current for my needs, so I've been fishing around for some time looking at alternative stepper boards. That's why I'm pleased to see  the links @ langwadt provided. Those ST boards provide quite a lot of punch for the price.

Now getting TinyG2 code running on STM32 M4  hooked up to some of those boards would fit the bill nicely.

**Comment by aldenhart** [comment]
I think you will find the TinyG code quite easy to work with. It's very well organized and commented. Please take a look. Let's play nice here.

**Comment by bronek999** [comment]
My problem was (and still is) with Adaptive Multi-Axis Step-Smoothing(AMASS).
I don't understund how it works :(.....  My problem was with generating(computing) delay cycles count.
Acceleration wasn't accelaration but set of random pulses. I found some problems with Keil ARM compiller, so I tried rewrite it to GNU C.

Now I found it support safety doors switch (v0.9i)....One of thinks I'm looking for.....

**Comment by chamnit** [comment]
@bronek999 : You can disable AMASS via a config.h define and the ifdef's in the stepper.c code should tell you what you can remove. That should help simplify porting. As for describing how AMASS works, I've been meaning to write something up about that, but it's more than an issues post should be.

**Comment by J-Dunn** [comment]
"TinyG2 is arduino style code. I dont' like it. "  That was my impression too. That is one reason why I thought it looked like it might be a nightmare to debug. 

The other reason is that it looked like some pretty cleaver coding but not at all easy to read. Sorry Aldan, it's not a case of playing nice. I did spend quite some time reading through it and found it very opaque and variable and function naming completely confusing and impossible to remember. ( Clear variable names do not require memory, they are clear each time they are read ). 

Now again, don't get the impression I'm trashing TinyG, it actually looks technically very impressive but my feeling was that I would spend more time trying to understand it to a level where I could maintain / debug it , than I would writing my own code. 

This is often the case and is an awkward trade-off. 

Now I'm sure if I just buy a board an use it stock, it a very good product. Hats off ! ;)

**Comment by J-Dunn** [comment]
What appeals about what chamnit has done is that he's seems to have ditched the awful Arduino libraries and written something clean and sensible.

**Comment by aldenhart** [comment]
@J-Dunn - Thanks for the feedback - style is a personal matter. 

Just as a point of clarification - There are  no Arduino libraries in the code (never have been). The C++ code is written on top of a hardware abstraction layer (Motate) that enables a high degree of platform portability without sacrificing performance.

**Comment by csdexter** [comment]
Probably, that's what they meant when they said "Arduino-style code" :-) They meant C++ which, to somebody coming from a world in which you only see C and ASM, is kinda scary.

**Comment by aldenhart** [comment]
Fair enough. The Motate underpinnings are C++ - the upper layers are prettymuch straight C code

**Comment by csdexter** [comment]
Also, on the S-curve thing, I'm working on the same thing right now and it's not obvious. Now, I know my [math-]intelligence level is not something extraordinary but even so, the amount of stuff that has to get in to get a proper end-to-end pipeline that can read movement commands and output pulses -- using S-curves in the middle -- is far from trivial.

**Comment by aldenhart** [comment]
Indeed. It's a rather deep well.

**Comment by bronek999** [comment]
I copied required file again...
Project for Keil based on GRBL v0.8 www.amega.sk/cnc/cnc.zip - it is full working
Test project for porting GRBL v0.9 www.amega.sk/cnc/stm32-test.zip - very first alpha... - not working correctly
It is GNU C project to eclipse.

I had problem in - uint32_t cycles = ceil( (TICKS_PER_MICROSECOND*1000000*60)*inv_rate ); 
file stepper.c procedure void st_prep_buffer().

**Comment by cri-s** [comment]
This code is something not portable.
 (TICKS_PER_MICROSECOND*1000000*60)
Limit is 71Mhz, then it's over the 32bit range.
Inv_rate is something like 0.000400

It's better to do it:
 inv_rate*=TICKS_PER_MICROSECOND*1000;
 uint32_t cycle=inv_rate*60000;
 if(inv_rate!=cycles) cycles++; // ceil function

Because inv_rate is not used after, you can do it this way, just add
note and maybe use
#if 0
 uint32_t cycles = ceil( (TICKS_PER_MICROSECOND100000060)*inv_rate );
#else // workaround for portability, modifies inv_rate
 inv_rate*=TICKS_PER_MICROSECOND*1000;
 uint32_t cycle=inv_rate*60000;
 if(inv_rate!=cycles) cycles++; // ceil function
#endif


#endif
Further

2015-03-19 10:26 GMT+01:00, bronek999 <notifications@github.com>:

**Comment by J-Dunn** [comment]
langwadt commented Nov 22, 2014
</blockquote>
1 * http://www.st.com/web/catalog/tools/FM116/SC959/SS1532/LN1847/PF260320
3 * http://www.st.com/web/en/catalog/tools/PF260715

~$40 for 3 axis with 3Amp drivers, 100MHz Cortex M4

All you need to do is move a few resistors and stack the boards, no messing with potmeters or jumpers to set microstepping or currrent, it is all done via SPI</blockquote>

This looks like a very interesting prospect. since it's arduino compatible pinouts it should work on top of  grbl on a Uno.  

I want to run a X-axis driven on both sides, so stacking two of those boards with same resistor setting should provide that functionality.

With  bronek999's code (thanks)  we have at least v0.8 ported to stm32. It should just be a case of pin configurations to set up for the nucleo board.

Presumably a little further work could finish porting v0.9.  
8)

**Comment by bronek999** [comment]
My "port" is no 100% compatible with grbl. It is project based on grbl without uart and with touch lcd and sd card support.


Odoslané zo služby Samsung Mobile

-------- Pôvodná správa --------
Od: J-Dunn <notifications@github.com> 
Dátum:2015. 03. 20  20:31  (GMT+01:00) 
Komu: grbl/grbl <grbl@noreply.github.com> 
Kópia: bronek999 <bronek@amega.sk> 
Predmet: Re: [grbl] New hardware suggestion teensy3.1 (#346) 

langwadt commented Nov 22, 2014

1 * http://www.st.com/web/catalog/tools/FM116/SC959/SS1532/LN1847/PF260320
3 * http://www.st.com/web/en/catalog/tools/PF260715

~$40 for 3 axis with 3Amp drivers, 100MHz Cortex M4

All you need to do is move a few resistors and stack the boards, no messing with potmeters or jumpers to set microstepping or currrent, it is all done via SPI

This looks like a very interesting prospect. since it's arduino compatible pinouts it should work on top of grbl on a Uno.

I want to run a X-axis driven on both sides, so stacking two of those boards with same resistor setting should provide that functionality.

With bronek999's code (thanks) we have at least v0.8 ported to stm32. It should just be a case of pin configurations to set up for the nucleo board.

Presumably a little further work could finish porting v0.9.

8)

**Comment by J-Dunn** [comment]
thanks for the clarifications. 

touch screen and SD is the way I want to go once basics are working. By without UART do you mean that you take the gcode from SD instead of UASB?

**Comment by bronek999** [comment]
Yes. My usb is for 5v power only,but add uart support require add init,interrupt handling. I'm using one uart (rs485) for spindle control. 


Odoslané zo služby Samsung Mobile

-------- Pôvodná správa --------
Od: J-Dunn <notifications@github.com> 
Dátum:2015. 03. 20  20:45  (GMT+01:00) 
Komu: grbl/grbl <grbl@noreply.github.com> 
Kópia: bronek999 <bronek@amega.sk> 
Predmet: Re: [grbl] New hardware suggestion teensy3.1 (#346) 

thanks for the clarifications.

touch screen and SD is the way I want to go once basics are working. By without UART do you mean that you take the gcode from SD instead of UASB?

**Comment by J-Dunn** [comment]
Great, so you are able to run motors using grbl at the same time as maintaining real time output on the screen?  Up to what kinds of speeds and does this work reliably?

This is very interesting.

**Comment by bronek999** [comment]
"basic" graphics operation runs on "main" program. grbl is running by interrupt. So it is independent. Only actual possition pixel is painted by timer interrupt. It have one disatvantage- I can't touch LCD monitor while SD card reading. It uses one SPI, file reading is in main procedure and touch detect is in interrupt. Now I have another pcb with touch panel by SPI and SD is by SDIO.

**Comment by J-Dunn** [comment]
Thanks, that is what I would expect. I think this sort of mix of periferals needs to be built on top of some kind of RTOS like Chibios.

**Comment by J-Dunn** [comment]
I've bought a stack of X-NUCLEO-IHM01A1 boards as linked above. They are basically a breakout board for the very capable L6474 chip. However, most of the set up and control for that chip is via SPI. 

Ultimately I intend to port GRBL to STM32-M4 and I already have an ST board for that, but in order to avoid doing everything in one jump as created a huge hardward+software debugging headache, I would like to take advantage of the Arduino pin format of these boards to get at least one motor running on the bench with fairly standard GRBL v0.9

Is there enough room left to add basic SPI capabilities in order to program the L6474, or is it already shoehorned in so tight it's not even worth trying?

Thanks for any tips that may save me some time.

**Comment by langwadt** [comment]
afaict the st board is not directly compatible with GRBL, the step/dir pins are spread over two ports so you'll have to write a bit of code to set the pins, but you'll have to do that for the STM port anyway

**Comment by J-Dunn** [comment]
Thanks. Yes if realise it needs some fiddling mainly to free up the SPI port which is used by the limit switch inputs. 

I was a bit confused for a while by the "PINOUT" pins which to be inputs !  My understanding of 'pinout' is a pin layout for a chip and has nothing to do with it being input or output, so you can't have a 'pinout pin'. I don't know what the logic of this naming is but it's pretty confusing and does not really make sense. Maybe it means something else in Swedish :?

My game plan is to move the limit switches to ADC port 2,3,4 and leave z-probe at 5.  Reset and Feedhold stay on 0,1 and I lose Cycle_start. ( Don't think I need that anyway ). 

There are some comments in the code about not mixing these but they're all inputs and I think it just means testing in the first line of the pin change interrupt  ISR whether it is a limit sw input or one of "pinout" inputs which has changed. 

To run three of these SPI boards will need two more lines for additional CS pins. It looks like I will have to wire_OR the three limit switches.  If it's homing one axis at at time that will suffice. If any axis hits a limit sw during operation it's the same error response and a manual intervention. So that seems OK too. 

It's obviously a bit of a squeeze on the limited pinout of an Arduino Uno but I think it should be possible.

Do you see any problems with that scheme?

**Comment by J-Dunn** [comment]
There is also some talk of daisy-chaining these boards but it's very poorly documented and looks like a major headache to workout what it is supposed to do and implement it properly. It seems like an inefficient hack to avoid using two extra lines for individual CS pins. 

Since this is just a stepping stone to porting it to M4, that is something I can skip altogether, I think.

**Comment by chamnit** [comment]
@J-Dunn : I'm sorry to hear that the naming conventions are confusing. Some of them were written when I was blurry-eyed and have since stuck. A lot of them have recently been updated to be a little less confusing and more descriptive. Please point out or provide a pull-request on updating some of the naming conventions that I missed. I try to keep everything as well-commented and neatly written as possible for people like you who are reviewing the code for their own purposes. (FYI, I'm not Swedish or Norwegian. I'm a US-born citizen living in New Mexico).

As for daisy-chaining, it is poorly documented because it's not supported. It won't be directly supported until other higher priority items are installed. That said, the tools are all there. You just need to communicate to Grbl through the serial interface just like a GUI should. It's not much more complicated than that.

**Comment by J-Dunn** [comment]
" the step/dir pins are spread over two ports"

Huh? That's not what I'm seeing in cpu_map.h, it seems that all six lines are on portB. 

I hope Markdown dues not mess this up ...

> // Define step pulse output pins. NOTE: All step bit pins must be on the same port.
>  #define STEP_DDR        DDRD
>  #define STEP_PORT       PORTD
>  #define X_STEP_BIT      2  // Uno Digital Pin 2
>  #define Y_STEP_BIT      3  // Uno Digital Pin 3
>  #define Z_STEP_BIT      4  // Uno Digital Pin 4
>  #define STEP_MASK       ((1<<X_STEP_BIT)|(1<<Y_STEP_BIT)|(1<<Z_STEP_BIT)) // All step bits
>
>  // Define step direction output pins. NOTE: All direction pins must be on the same port.
>  #define DIRECTION_DDR     DDRD
>  #define DIRECTION_PORT    PORTD
>  #define X_DIRECTION_BIT   5  // Uno Digital Pin 5
>  #define Y_DIRECTION_BIT   6  // Uno Digital Pin 6
>  #define Z_DIRECTION_BIT   7  // Uno Digital Pin 7
>  #define DIRECTION_MASK    ((1<<X_DIRECTION_BIT)|(1<<Y_DIRECTION_BIT)|1<<Z_DIRECTION_BIT)) // All direction bits

**Comment by bronek999** [comment]
Today I have connected 7 inch LCD (SPI mode), SD card (SPI mode) and capacitive multi touch screen(TWI) to Arduino Due (SAM3X8E) running TinyG2. Hardware works, I can start write software.

**Comment by J-Dunn** [comment]
Thanks chamnit .  Firstly the daisy-chain thing was a reference to the ST boards, not GRBL, it may actually help fit these boards onto GRBL ( since it saves two lines ) if it was clearer what how it was supposed to work , but I'm not inclined to bang my head on that one. 

I did not think you were sweedish but I thought the original author was and guessed that these names may have a long history.  I hesitated to make a suggestion for a better naming convention until I had worked more with the code and have a better understanding of what these pins are for. They seem to be for manual hardware inputs so it would be good to get rid of the "OUT" part. 

If they can be described as external inputs maybe EXT_INPUT_PORT , EXT_INPUT_MASK , etc or something similar. 

I much appreciate the way you have written this: simple, clear and efficient and you availability to comment and help here.  Just scanning the code I have a rough idea of how I need to handle this. Any comments where I'm fooling myself may be helpful.  I have not fully thought through the idea of mixing the extra CS lines into ADC port yet .. :?

**Comment by langwadt** [comment]
The daisy chaining is straight forward I use it in a different project. 
just connect output from one chip to input of the next, then instead of clocking out(in) 8bits and setting 
CS high, you clock out(in) n*8bits before setting CS high, the first 8bit out ends up in the n'th chip

The comment about pins being spread over two ports, was in relation to the ST boards having one of the step pins in portB and the rest in portD so it isn't directly compatible with GRBLs requirement for them being in one port

**Comment by J-Dunn** [comment]
Oh, so it acts like a shift register until the common CS goes high at which point each chip decodes what it has.  So can different things can be sent to each chip, or does it have to be the same thing in and out of each one?

Where is all this documented? I have only found the breifest mention in the L6474 datasheet.

I've just found this thing about where they put the CS and SCK pins, rather an odd choice and a bit inconvenient. since Arduino SPI is all on one port by default that would have been good. Maybe they have a technical reason to spread it around. 

Looks like another complication. :(

**Comment by J-Dunn** [comment]
It should be said that they intend this for use with their Nucleo boards, but adopting the Arduino pinout would have made it more logical to follow Arduino pin usage.

**Comment by langwadt** [comment]
yes it acts like a shift register, you can send different things to different chips. With a bit of reading between the lines it is all in the datasheet, there maybe an app. note I don't remember

I just bit bang the spi, not much that gained with a hardware spi anyway, usually you only need to set the current and possible read out errors when the FLAG gets set

**Comment by sebbra** [comment]
I have some X-NUCLEO-IHM01A1 too, and i will build a setup with a ATMEGA1284P to have some more Ram for the SPI implementation.

here is the app-note on how to use the daisychain: 
http://www.st.com/st-web-ui/static/active/en/resource/technical/document/user_manual/DM00156746.pdf

and there is an arduino excample:
http://forum.arduino.cc/index.php?topic=295928.0

**Comment by J-Dunn** [comment]
OK Michel_F , who has written an arduino library for these ST boards has helped me suss this daisy-chain thing. https://github.com/MotorDriver/L6474/issues/1#issuecomment-90471305

Basically the information required to use several chips is contained in a separate "application note" rather than in the chip's datasheet. Most manufacturers use application notes for worked examples not for essential chip specification data. ST seem to have other ideas.  There's also some incorrect information tables. 

Now I'm back to one CS pin , which will be less disruptive to grbl. ;)

They seem to control the step pins via PWM which is why they use both port B and D.  Hopefully in converting this to grbl I can just bend a couple of pins on the lowest board and jump it around. 

Time to roll my sleeves up on the grbl code ...

**Comment by J-Dunn** [comment]
chamnit : " A lot of them have recently been updated to be a little less confusing and more descriptive. Please point out or provide a pull-request on updating some of the naming conventions that I missed. "

since these pins seem to be external, manual inputs EXT_INPUT may be a clearer prefix. If that does not fit some actual or potential functionality there may be a better name. Getting away from calling an input by a name that suggests an output would be the main aim. 

The following seems to catch all instances:

>  sed -i 's/PINOUT/EXT_INPUT/g' cpu_map.h system.c

**Comment by J-Dunn** [comment]
Probably want to take care of comments too: 
> sed -i 's/pinouts/external input/g' cpu_map.h system.c
> sed -i 's/\(pinout\)//g' cpu_map.h system.c

**Comment by chamnit** [comment]
@J-Dunn : Thanks. I'll get this updated after I get done with this next feature push. Should be fairly soon.

**Comment by J-Dunn** [comment]
Cool. Is there new version imminent? I'll be pouring over the code for a while yet but I don't want to start porting stuff too much just to find I have to redo it all in a week's time ;)

> sed -i 's/pinout/external input/g' main.c

**Comment by chamnit** [comment]
@J-Dunn : No. This stuff is more OEM specific, but there has been enough refactoring to the code there to warrant a vetting period. There are also other things I need to get installed after this. In short, there won't be a new master release anytime soon.

**Comment by J-Dunn** [comment]
Thanks, I've been ploughing through Grbl in detail. It's very easy to work with, nice job.

>      // Set step pulse time. Ad hoc computation from oscilloscope. Uses two's complement.
>      st.step_pulse_time = -(((settings.pulse_microseconds-2)*TICKS_PER_MICROSECOND) >> 3);

I take it the shift 3 corresponds to the default fastest time being CPU/8 . I see that the 2us is an empirical tweak, where is the that time lost, is that the ISR overhead on Uno?!

save a few bytes, and cycles in ISR:
>        // Initialize Bresenham line and distance counters
>        st.counter_z = st.counter_y = st.counter_x = (st.exec_block->step_event_count >> 1);

Some more explicit vars again may helpful, I'll make a list. 
>  Timer1_ISR_busy = true;

 sed -i 's/busy/Timer1_ISR_busy /g' cpu_map.h stepper.c
 sed -i 's/step_outbits/step_port_bits/g' cpu_map.h stepper.c

**Comment by chamnit** [comment]
@J-Dunn : THanks.

- `step_pulse_time` was measured on an oscilloscope. Yes, 2us is ISR overhead.
-  I'll add your suggestions on the next push.

**Comment by csdexter** [comment]
From the ATmega328P datasheet: an interrupt takes 4+3 clock cycles and returning from one takes 4 cycles. At 16MHz, one clock cycle is 62.5ns so 7 of them will be 437.5ns. This leaves 1562.5ns (25 clock cycles) as, I'm guessing, overhead introduced by GCC as it's pushing stuff on the stack to prepare the ISR context for execution.

**Comment by cri-s** [comment]
Times ago i have made some tests, source modified and as i remember,
plain unmodified code
takes over 5uS less on timing.
If you want, i can make timing for actual firmware, please indicate
what interest you.

cheers chris.

**Comment by J-Dunn** [comment]
There's about 14 registers that need to be pushed and then popped; stack pointer pushed; a few extras; reti etc. That is going to be close to 32 clock cycles.  I was a bit surprised when I saw 2us, but it's a slow chip and after counting clock cycles it like 2us is pretty close. 

Arduino library is full of cruft and probably takes twice as long to do the same thing, very inefficiently.

**Comment by chamnit** [comment]
@J-Dunn : I'm not exactly sure where the inefficiencies are, but Grbl doesn't use any of the Arduino IDE libraries (or any other libraries except than the standard AVR libs). Grbl only uses the GCC compilers and tools that the IDE provides. 

That said, I've never bothered to look at the assembly to see what exactly the stepper ISR is doing when it gets invoked and the ISR is doing its thing. The core design of the stepper ISR hasn't changed since I started working the Grbl project. If you all have any other insight on how it could be done differently, more efficiently, or know of better "good practices", I'm all ears.

**Comment by J-Dunn** [comment]
There may be a few tweaks that can be done but you misunderstood my comment. Grbl is probably close to as fast as it can be on that chip. What I meant was using Arduino libs ( which you do not ;) ) would make it about twice as slow as it actually is.  IIRC using the libs provided with the non_IDE blinking a LED takes almost 4us.

**Comment by KristianDIY** [comment]
Is there any version or port out that works on due yet? I thought I saw online that it would work as is... but after creating a terminal board and soldering everything to a headerless due, I could not get it working, now I know why, as the published version is for avr not sam, but is there one out there, or is there one in the making?
